src/basic/terminal-util.c

```diff
@@ -2427,6 +2430,125 @@ int terminal_get_size_by_dsr(
+        int r;
+
+        assert(buf);
+        assert(ret_name);
+
+        /* Check if the have enough space for an answer */
+        if (size < STRLEN(DCS_TERMINFO_R0))
+                return 0;
+
+        /* Check if the terminating sequence is present */
+        if (memcmp(buf + size - STRLEN(ANSI_ST), ANSI_ST, STRLEN(ANSI_ST)) != 0)
+                return 0;
+
+        if (size <= STRLEN(DCS_TERMINFO_R1 ANSI_ST))
+                return -EINVAL;  /* The answer is invalid or empty */
+
+        if (memcmp(buf, DCS_TERMINFO_R1, STRLEN(DCS_TERMINFO_R1)) != 0)
+                return -EINVAL;  /* The answer is not valid */
+
+        _cleanup_free_ void *dec = NULL;
+        size_t dec_size;
+        r = unhexmem_full(buf + STRLEN(DCS_TERMINFO_R1), size - STRLEN(DCS_TERMINFO_R1 ANSI_ST),
+                          /* secure= */ false,
+                          &dec, &dec_size);
+        if (r < 0)
+                return r;
+
+        assert(((const char *) dec)[dec_size] == '\0'); /* unhexmem appends NUL for our convenience */
+        if (memchr(dec, '\0', dec_size) || string_has_cc(dec, NULL) || !filename_is_valid(dec))
+                return -EUCLEAN;
+
+        *ret_name = TAKE_PTR(dec);
+        return 1;
+}
+
+int terminal_get_terminfo_by_dcs(int fd, char **ret_name) {
+        int r;
+
+        assert(fd >= 0);
+        assert(ret_name);
+
+        /* Note: fd must be in non-blocking read-write mode! */
+
+        struct termios old_termios;
+        if (tcgetattr(fd, &old_termios) < 0)
+                return -errno;
+
+        struct termios new_termios = old_termios;
+        termios_disable_echo(&new_termios);
+
+        if (tcsetattr(fd, TCSADRAIN, &new_termios) < 0)
+                return -errno;
+
+        r = loop_write(fd, DCS_TERMINFO_Q, SIZE_MAX);
+        if (r < 0)
+                goto finish;
+
+        usec_t end = usec_add(now(CLOCK_MONOTONIC), CONSOLE_REPLY_WAIT_USEC);
+        char buf[STRLEN(DCS_TERMINFO_R1) + MAX_TERMINFO_LENGTH] = {};
+        size_t bytes = 0;
+
+        for (bool first = true;; first = false) {
+                usec_t n = now(CLOCK_MONOTONIC);
+                if (n >= end) {
+                        r = -EOPNOTSUPP;
+                        break;
+                }
+
+                r = fd_wait_for_event(fd, POLLIN, usec_sub_unsigned(end, n));
+                if (r < 0)
+                        break;
+                if (r == 0) {
+                        r = -EOPNOTSUPP;
+                        break;
+                }
+
+                /* On the first try, read multiple characters, i.e. the shortest valid reply. Afterwards read
+                 * byte by byte, since we don't want to read too much and drop characters from the input
+                 * queue. */
+                ssize_t l = read(fd, buf + bytes, first ? STRLEN(DCS_TERMINFO_R0) : 1);
+                if (l < 0) {
+                        if (errno == EAGAIN)
+                                continue;
```

Comment:
When `first` is true and `EAGAIN` is received, then we will read one by one. Is it expected?

Reply:
Yeah, I copied this pattern from the other functions. I changed the condition to `bytes == 0`.